### PR TITLE
[WIP] Change PoW to Equihash

### PR DESCRIPTION
**Pull request by h4x3rotab**
This ongoing PR changes the PoW to Equihash. The implementation of Equihash is taken from Zcash project.

It will include the following modifications:

- [x] Link sodium for Equihash library
- [x] Extend block header to include Equihash solution
    - [x] Embed Equihash related consensus parameters
    - [x] Introduce a new block version
    - [x] Add Equihash solution field to block header
    - [x] Legacy node serialization compatibility
    - [x] Be Zcash miner compatible
- [x] Change PoW after fork height
    - [x] Add PoW functions
    - [x] Introduce Equihash block validation for the hard fork
    - [x] Fix mining RPCs
- [x] Fix rest RPCs
- [ ] Tests
    - [ ] Unittest everywhere
    - [ ] Regression test
- [ ] Fix compact block

### CODE REVIEW

**Comment by h4x3rotab** [comment]
Compact block is broken by new block format. Seems we have to fix it.

**Comment by sennevb** [comment]
Watching this one closely, interested to add this coin to the pool on launch..

**Comment by daira** [comment]
(Zcash developer here.)

Zcash will also need to add compact blocks. We haven't thought about it much yet but our ticket for that is https://github.com/zcash/zcash/issues/2316 . Let's cooperate on a spec.

File: src/net.h [hunk 1]
```diff
@@ -833,6 +830,8 @@ class CNode
     std::string GetAddrName() const;
     //! Sets the addrName only if it was not previously set
     void MaybeSetAddrName(const std::string& addrNameIn);
+
+    bool IsLeacyBlockHeader(int version) { return version < BTG_HARD_FORK_VERSION; };
```
Comment (daira) [inline src/net.h]: Typo in function name.

Comment (h4x3rotab) [reply src/net.h]: Fixed. Thanks

**Comment by antonifigueiredo** [comment]
Boa tarde,  qual a previsão de Exchange terá o BTG? (BITCOINGOLD)?

**Comment by h4x3rotab** [comment]
It's time to do mining tests. No way to do it in a regular regression test as there's no existing Equihash solver in Python.

**Comment by peter-bonanni** [comment]
@h4x3rotab 

https://github.com/str4d/zcash-pow

Isn't this an Equihash python solver?

**Comment by oliverw** [comment]
@h4x3rotab Is the daemon ready for pool integration now?

**Comment by h4x3rotab** [comment]
This change should be further tested but for now let's merge it to the staging tree. I will create another issue to track the process of PoW changes & tests.

**Pull request merged**